blog/2024-09-02-amqp-flow-control/index.md

```diff
@@ -0,0 +1,762 @@
+A link connects either a sending client application to an [exchange](/tutorials/amqp-concepts#exchanges) in RabbitMQ or a [queue](/tutorials/amqp-concepts#queues) in RabbitMQ to a consuming client application.
+
+:::info
+
+While the AMQP 1.0 specification uses the terms "senders" and "receivers", the RabbitMQ documentation refers to ["publishers"](/docs/publishers) (or "producers") and ["consumers"](/docs/consumers).
+When discussing client applications, these terms can be used interchangeably. Therefore, a client application instance that:
+* Sends messages to RabbitMQ is a **sender** / **publisher** / **producer** (with RabbitMQ acting as the **receiver**).
+* Receives messages from RabbitMQ is a **receiver** / **consumer** (with RabbitMQ acting as the **sender**).
+
+:::
+
+### Link-Credit
+
+The central idea behind AMQP 1.0 link flow control is simple:
+**To receive messages, a consumer must grant credits to the sending queue.**
+
+One credit corresponds to one message.
+For example, when the consumer grants 10 credits, RabbitMQ is allowed to send 10 messages.
+This straightforward principle, where the receiver provides **feedback** to the sender, ensures that the sender never overwhelms the receiver.
+
+Both the receiver and sender maintain their own "link state".
+Part of this state is the current link credit.
+Link credit decreases by 1 each time a message is transferred.
+Specifically, the sender reduces link credit by 1 when it sends a message, and the receiver reduces link credit by 1 when it receives a message.
+When the sender's link credit reaches 0, it must stop sending messages.
+
+Messages are sent in [transfer](https://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transport-v1.0-os.html#type-transfer) frames.
+
+Credits are granted in [flow](https://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transport-v1.0-os.html#type-flow) frames:
+```xml
+<field name="link-credit" type="uint"/>
+```
+As you might have guessed, they are called "flow" frames because these frames carry flow control information.
+The type [uint](https://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-types-v1.0-os.html#type-uint) stands for unsigned integer, a value between 0 and a large number (2^32 - 1).
+
+Even after the link is successfully set up ("[attached](https://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transport-v1.0-os.html#doc-idp315568)" in AMQP 1.0 terms), RabbitMQ is not allowed to start sending messages to the consumer until the consumer sends its first `flow` frame, granting link credit to the sending queue.
+
+In its simplest form, when a client (receiver) grants a single credit to the queue (sender), the queue will send a single message, as illustrated in [Figure 2.43](https://docs.oasis-open.org/amqp/core/v1.0/os/amqp-core-transport-v1.0-os.html#doc-idp416352) of the AMQP 1.0 specification:
+
+![AMQP 1.0 Figure 2.43: Synchronous Get](figure-2-43.png)
```

Comment:
Is it possible to use text instead of a screenshot here?

Reply:
Yes, text is indeed better, thank you!